### PR TITLE
WIP: Add ability to record LFP's

### LINKED ISSUE #68: add CSD plot for comparing against animal studies
### Background

The ability to calculate LFP for simulated electrodes in the model (#150) allow for time vs. depth laminar CSD plots to be generated. Laminar CSD (electrodes arranged vertically) is the primary goal of this feature because laminar CSD is commonly used in animal studies where a laminar probe is inserted into the cortext to measure extracellular potentials. In the experimental animal studies, CSD is calculated from the bandpass filtered LFP data (low-frequency portion of extracellular potentials) to identify underlying neuronal activity at specific cortical layers. Such activity is identified as a "sink" (net influx of current into cell) or "source" (efflux into extracellular space) in the temporal-spatial domain.

Figure 5 from Sherman et al. 2016 shows laminar CSD plots of an anesthetized mouse and an awake monkey.

Screen Shot 2020-09-08 at 8 41 05 AM

> Sherman MA, Lee S, Law R, et al. Neural mechanisms of transient neocortical beta rhythms: Converging evidence from humans, computational modeling, monkeys, and mice. Proc Natl Acad Sci U S A. 2016;113(33):E4885-E4894. doi:10.1073/pnas.1604135113

CSD can be calculated via
```
numpy.diff(LFP_data,n=2,axis=ax)/spacing_mm**2
```

The sign may be adjusted to make a current sinks negative and sources positive.

### Goal

Create laminar CSD plots from simulated data that are comparable to Figure 5 in Sherman et al. 2016

### Requirements

1. Averaging LFP over multiple trials is typically done before computing CSD
2. CSD plots are created using a 2D spline over values at each electrode depth. Support various spline functions.
3. Users need to be able to easily adjust the X,Y scales to match simulated data with their experimental data
4. The CSD of the first and last channels cannot be computed using the 2nd spatial derivative. The Vaknin method is used to derive approximate CSD signals at these points

    > G. Vaknin, P.G. DiScenna, T.J. Teyler, A method for calculating current source density (CSD) analysis without resorting to recording sites outside the sampling volume, Journal of Neuroscience Methods, Volume 24, Issue 2, 1988, Pages 131-135, ISSN 0165-0270, https://doi.org/10.1016/0165-0270(88)90056-8.
5. Implement [iCSD method](https://github.com/espenhgn/iCSD) as an alternative to `numpy.diff()` above

### PR DESCRIPTION
**Pull request by jasmainak**
cc @blakecaldwell 

is this the right file?

closes #68 

I just copied it for now. I'll work on it to improve and understand it.

### CODE REVIEW

**Comment by blakecaldwell** [comment]
Yes, this is the right file for LFP. Calculating CSD is here too. However, plotting CSD (time vs. depth) is kind of complicated and will probably look very different. Perhaps this PR can just be for LFP.

**Comment by jasmainak** [comment]
okay, I improved the implementation of PSA and LSA here. Code looks much cleaner if you use numpy.

@blakecaldwell do we have some ground truth? I don't know if we're getting the right result when I run the code.

File: hnn_core/lfp.py [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+        if pc is None:
+            self.pc = h.ParallelContext()
+        else:
+            self.pc = pc
+
+    def setup(self):
+        """Enables fast calculation of transmembrane current (nA) at
+           each segment."""
+        h.cvode.use_fast_imem(1)
+        self.bscallback = h.cvode.extra_scatter_gather(0, self.callback)
```
Comment (jasmainak) [inline hnn_core/lfp.py:91]: I had to replace the line [here](https://github.com/jonescompneurolab/hnn/blob/master/lfp.py#L117) with this because my version of `Neuron` doesn't have the function. Maybe @samnemo do you know what's the right thing to do here?

**Comment by jasmainak** [comment]
Here is the plot of the LFP for the two methods

Screen Shot 2020-09-01 at 12 14 41 AM

they match exactly. Does it look correct to people? If so, I'll go ahead and add tests and example

Feel free to pull the branch and play with it.

**Comment by blakecaldwell** [comment]
> they match exactly. Does it look correct to people? If so, I'll go ahead and add tests and example

The correctness of these depends on the coordinates of the electrode and cells around it (i.e. the LFP above and below the soma will be drastically different, opposite I think). Probably the best way to verify correctness is to calculate LFP at various points around a SINGLE neuron. Many Einevoll papers did this and it'd be excellent to do the same with our code. This is large enough to be a separate pull issue (perhaps a hackathon issue).

**Comment by ntolley** [comment]
For validating the "biophysical intuition" a plot like this may be useful:
image

File: hnn_core/lfp.py (hunk 1 above)
Comment (blakecaldwell) [reply hnn_core/lfp.py:88]: @jasmainak I'm guessing you saw this post?
https://www.neuron.yale.edu/phpBB/viewtopic.php?p=15713#p15713

`cvode.extra_scatter_gather` seems to replace the `cvode.beforestep_py` functionality? Anyway, I can't find it in the neuron documentation anymore.

Comment (jasmainak) [reply hnn_core/lfp.py:88]: yeah, I figured from the API documentation that this must be the function :)

**Comment by jasmainak** [comment]
@ntolley looks like you had a very productive hackathon. I just added the code as a commit here so as to not lose it. Would love to learn from you what you did. Any way we can add some print statements so I know how far the simulation went?

**Comment by ntolley** [comment]
I can see about implementing that. Just to make sure I understand, do you mean print statements similar to the `simulate()` function in `parallel_backends.py`?

**Comment by jasmainak** [comment]
yep ... something similar :) You can push straight to this branch. My patience runs out after a script runs for 30 seconds, specially if I see no output.

Also before we merge, do make sure to remind me to credit everyone involved in writing the code.

**Comment by jasmainak** [comment]
@ntolley let's try to revive this PR once you are done with #276

**Comment by ntolley** [comment]
Read my mind, I was actually thinking about this PR today as well. We can discuss later, but I think the most straightforward implementation will be passing a list of tuples containing xyz coordinates. Something like:
```py
electrode_list = [(x1, y1, z1), (x2, y2, z2), ...]
net.record_lfp(electrode_list)
```

**Comment by jasmainak** [comment]
That sounds good! Some of the code here which takes a long time to run can be moved to demo notebooks/gists elsewhere.

**Comment by ntolley** [comment]
Very rough initial pass at this. Currently electrode positions are specified as a list of tuples under `net.lfp_electrode_pos`. After running a simulation the output is stored under `net.lfp`:
image

While the `Network` class is getting pretty bloated, this seems like a natural place to store electrode information at the moment. Absolutely no idea if this is being calculated correctly, we'll need to think of some good tests to validate the code.

**Comment by jasmainak** [comment]
Agree about storing it in `net`. However, you might want to "reset" it before the next simulation. See discussion with @cjayb and @blakecaldwell about simulation times

File: examples/plot_simulate_lfp.py [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+ns.start = 100
+ns.interval = 50.0
+ns.noise = 0.  # deterministic
+
+nc = h.NetCon(ns, cell.synapses['apicaltuft_ampa'])
+nc.weight[0] = 0.001
+
+h.tstop = 500.0
+lfp_rec = {'lsa': ([], []), 'psa': ([], [])}
+for method in ['lsa', 'psa']:
```
Comment (jasmainak) [inline examples/plot_simulate_lfp.py:69]: I think this script could go into @cjayb 's repo: https://github.com/cjayb/hnn-core-examples ? Because it runs pretty slow ... could be a nice place to host some of the longer examples

File: hnn_core/lfp.py [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+    nc.weight[0] = 0.001
+
+    h.tstop = 2000.0
+
+    for method in ['lsa', 'psa']:
+        elec = LFPElectrode([0, 100.0, 100.0], pc=h.ParallelContext(),
+                            method='psa')
+        elec.setup()
+        elec.LFPinit()
+        h.run()
```
Comment (jasmainak) [inline hnn_core/lfp.py:235]: This could be factored into a unit test. If you read the reference paper, I think you might find some corner cases in which the LSA and PSA method give the exact same result ... (I suspect)

**Comment by jasmainak** [comment]
How about renaming [pos_dict](https://github.com/jonescompneurolab/hnn-core/blob/master/hnn_core/network.py#L175) to be `pos_cells` (slight misnomer but I can live with it ...) and then calling the LFP position `pos_lfp`. That way it's more [discoverable](https://github.com/jonescompneurolab/hnn-core/blob/master/hnn_core/network.py#L175). In Ipython, when you type `net.pos` and press tab, you'll see both options.

**Comment by ntolley** [comment]
Love the `net.pos_lfp` suggestion! 

Also realized an interesting mismatch, the `times` recorded for the dipole is one element longer than the LFP recordings (see [L55](https://github.com/jasmainak/hnn-core/blob/1de1c26de5f94b310a42daeeb0c16056d7db0dea/examples/plot_simulate_lfp.py#L55)) I believe it has to do with the way the LFP is calculated. Anyone against just padding the copying the first element's value twice?

**Comment by jasmainak** [comment]
@ntolley you really are getting good at this stuff! Let me know when it's ready to review!

File: examples/plot_simulate_lfp.py [hunk 4]
```diff
@@ -0,0 +1,58 @@
+"""
+=====================
+08. Plot Simulate LFP
+=====================
+
+This example demonstrates how to record local field potentials (LFPs).
+"""
+
+# Author: Nick Tolley <nicholas_tolley@brown.edu>
```
Comment (jasmainak) [inline examples/plot_simulate_lfp.py:9]: you can use #sphinx-gallery-thumbnail to point to the second or third image as the thumbnail (see the other examples). I think either of those are more meaningful

File: hnn_core/network_builder.py [hunk 5]
```diff
@@ -318,6 +323,7 @@ def _build(self):
 
         self.move_cells_to_pos()  # position cells in 2D grid
         self._connect_celltypes()
+        self._record_lfp()
```
Comment (jasmainak) [inline hnn_core/network_builder.py:326]: move it next to `_record_spikes` ?

File: examples/plot_simulate_lfp.py [hunk 6]
```diff
@@ -0,0 +1,58 @@
+"""
+=====================
+08. Plot Simulate LFP
```
Comment (jasmainak) [inline examples/plot_simulate_lfp.py:3]:
```suggestion
08. Record LFP
```

?

File: hnn_core/externals/lfp.py [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+import matplotlib.pyplot as plt
+
+
+def get_all_sections(sec_type='Pyr'):
+    ls = h.allsec()
+    ls = [s for s in ls if sec_type in s.name()]
+    return ls
+
+
+class LFPElectrode:
```
Comment (jasmainak) [inline hnn_core/externals/lfp.py:40]: private class for now? Since we don't expose it to user

File: hnn_core/externals/lfp.py [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+    """LFP electrode class.
+
+    Parameters
+    ----------
+    coord : tuple
+        The (x, y, z) coordinates of the LFP electrode.
+    sigma : float
```
Comment (jasmainak) [inline hnn_core/externals/lfp.py:47]: I think folks might want access to this parameter. Any way we can expose it? Wouldn't hurt to expose "method" as well.

File: hnn_core/externals/lfp.py [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+            # for seg in s # so do not need to use segments...?
+            # more accurate to use segments and their neighbors
+            self.imem_ptrvec.pset(i, seg._ref_i_membrane_)
+
+        self.vres = self.transfer_resistance(self.coord, method=self.method)
+        self.lfp_t = h.Vector()
+        self.lfp_v = h.Vector()
+
+    def callback(self):
+        # print('In lfp callback - pc.id = ',self.pc.id(),' t=',self.pc.t(0))
```
Comment (jasmainak) [inline hnn_core/externals/lfp.py:190]:
```suggestion
```

File: hnn_core/externals/lfp.py [hunk 10]
```diff
@@ -0,0 +1,206 @@
...
+                # phi
+                num = np.sqrt(H ** 2 + r_sq) - H
+                denom = np.sqrt(H ** 2 + r_sq) - L
+                phi = 1. / dis * np.log(num / denom)
+
+            # x10000 for units of microV : nA/(microm*(mS/cm)) -> microV
+            vres.append(10000.0 * phi / (4.0 * np.pi * sigma))
+        return vres
+
+    def LFPinit(self):
```
Comment (jasmainak) [inline hnn_core/externals/lfp.py:173]: method name should be in camel case

File: hnn_core/network_builder.py [hunk 11]
```diff
@@ -459,6 +465,14 @@ def _connect_celltypes(self):
                     src_gid, nc_dict, target_cell.synapses[syn_key])
                 self.ncs[connection_name].append(nc)
 
+    def _record_lfp(self):
+        method = 'psa'
+        for pos in self.net.pos_lfp:
+            elec = LFPElectrode(pos, pc=_PC, cvode=_CVODE, method=method)
+            elec.setup()
+            elec.LFPinit()
```
Comment (jasmainak) [inline hnn_core/network_builder.py:473]: I think they should be one method

File: hnn_core/network.py [hunk 12]
```diff
@@ -1036,6 +1040,24 @@ def clear_connectivity(self):
         """Remove all connections defined in Network.connectivity_list"""
         self.connectivity = list()
 
+    def add_electrode(self, electrode_pos):
+        """Specify coordinates of electrodes for LFP recording.
+        Parameters
```
Comment (jasmainak) [inline hnn_core/network.py:1045]:
```suggestion

        Parameters
```

**Comment by jasmainak** [comment]
Pushed a clean up commit

File: hnn_core/externals/lfp.py [hunk 13]
```diff
@@ -0,0 +1,193 @@
+"""
```
Comment (jasmainak) [inline hnn_core/externals/lfp.py:1]: This file shouldn't be in externals. You keep stuff in externals when it needs to be synced with another repository. We have made extensive changes to the original code

**Comment by ntolley** [comment]
Decided to consolidate the attributes under `net.lfp`. It is now a list of dictionaries which contain information for each electrode, as well as the recording after the simulation is run. Here's an example from the tutorial:
```py
[{'lfp': [], 'pos': (2, 2, 400), 'sigma': 3.0, 'method': 'psa'},
 {'lfp': [], 'pos': (6, 6, 800), 'sigma': 3.0, 'method': 'psa'}]
```

I think this storage method has the smallest potential for confusion w.r.t. where each recording comes from. Open to alternative suggestions though.

File: hnn_core/network_builder.py (hunk 5 above)
Comment (ntolley) [reply hnn_core/network_builder.py:326]: It's actually really important this is placed after `self.move_cells_to_pos()`. When `LFPElectrode` is instantiated, it keeps track of where each cell is positioned to perform the voltage calculation. 

Do you think it's worth adding a comment to make sure people don't move it in the future?

**Comment by jasmainak** [comment]
I just have a tiny suggestion. I'd call it 'data' instead of 'lfp' because it's already `net.lfp`

File: examples/plot_simulate_lfp.py [hunk 14]
```diff
@@ -0,0 +1,64 @@
+"""
+=====================
+08. Record LFP
+=====================
```
Comment (jasmainak) [inline examples/plot_simulate_lfp.py:4]: align with text

File: hnn_core/network_builder.py (hunk 5 above)
Comment (jasmainak) [reply hnn_core/network_builder.py:327]: yes okay. Best would have been to have a check that if `pos_dict` is not created, then fail or something of that sort

**Comment by jasmainak** [comment]
Time to write tests?!

File: hnn_core/parallel_backends.py [hunk 15]
```diff
@@ -65,6 +65,9 @@ def _gather_trial_data(sim_data, net, n_trials, postproc):
         net.cell_response.update_types(net.gid_ranges)
         net.cell_response._vsoma.append(spikedata[3])
         net.cell_response._isoma.append(spikedata[4])
+        assert len(net.lfp) == len(spikedata[5])
```
Comment (blakecaldwell) [inline hnn_core/parallel_backends.py:68]: can this assert be removed?

**Comment by blakecaldwell** [comment]
thanks @ntolley!

File: hnn_core/parallel_backends.py (hunk 15 above)
Comment (ntolley) [reply hnn_core/parallel_backends.py:68]: I suppose that's better suited in the testing code. The main motivation for this check is to make sure recordings aggregated during MPI runs are formatted correctly.

File: hnn_core/network_builder.py (hunk 5 above)
Comment (ntolley) [reply hnn_core/network_builder.py:327]: I guess the problem is `pos_dict` will always exist once this point is reached (since it's made in the Network instantiation), and the cells will have a default position when they are created. 

Maybe one way around this would be to pass the cell position to `LFPElectrode` through `net.pos_dict`, but I'm not sure if there's a huge benefit.

Comment (jasmainak) [reply hnn_core/network_builder.py:327]: I guess what I'm arguing for is a PR to remove `network_builder.move_cells_to_pos`. The transformation should happen before you go into Neuron. Ideally [after this line](https://github.com/jonescompneurolab/hnn-core/blob/master/hnn_core/network.py#L240)

Comment (ntolley) [reply hnn_core/network_builder.py:327]: Good suggestion, it's honestly not clear to me why this doesn't just take place during `network_builder._create_cells_and_drives()`.

File: hnn_core/parallel_backends.py (hunk 15 above)
Comment (blakecaldwell) [reply hnn_core/parallel_backends.py:68]: yeah, assert statements in user code means a bug has gone without fixing. if an exception can be identified, it should be handled properly so as to not crash the code. they are useful in tests because they tell developers to fix the code (the user isn't expected to go into the source code).

File: hnn_core/tests/test_network.py [hunk 16]
```diff
@@ -212,6 +212,40 @@ def test_network():
     net.clear_connectivity()
     assert len(net.connectivity) == 0
 
+    # Test LFP electrodes
+    kwargs_default = {
+        'electrode_pos': (2, 2, 400),
+        'sigma': 3.0, 'method': 'psa'}
```
Comment (jasmainak) [inline hnn_core/tests/test_network.py:218]: I wouldn't recommend using this pattern actually. I think it somehow made sense in the last test. For many newcomers to Python, kwargs is a confusing concept. If you can just set the variables like:

```py
electrode_pos = (2, 2, 400)
sigma = 3.0
method = 'psa'
```

also I think you need to have a new file `test_lfp.py` where you test the actual LFP calculations itself. Harder but must be done :)

Comment (jasmainak) [reply hnn_core/tests/test_network.py:218]: Maybe spend some time reading some of the tests in MNE-Python. It might help you to write better tests. For example:

https://github.com/mne-tools/mne-python/blob/main/mne/tests/test_epochs.py

You'll see how there is other stuff tested along with the `ValueError`

File: hnn_core/parallel_backends.py (hunk 15 above)
Comment (jasmainak) [reply hnn_core/parallel_backends.py:68]: Absolutely this should go into the test

File: hnn_core/lfp.py (hunk 3 above)
Comment (ntolley) [reply hnn_core/lfp.py]: Is this the sort of setup you're thinking should go in `test_lfp,py` for testing the code directly?

Comment (jasmainak) [reply hnn_core/lfp.py]: yeah one test like that which checks that LSA and PSA give roughly equal results. I think they would be more equal the further you go away from the cell.

And then one where you could just look at the `vres` and see if it's computed correctly. I'm worried that [lines like this](https://github.com/jasmainak/hnn-core/blob/lfp/hnn_core/lfp.py#L105-L106) might be incorrect (see also [this figure](https://www.frontiersin.org/files/Articles/197337/fncom-10-00065-HTML/image_m/fncom-10-00065-g001.jpg)). You can give different positions and see if the `vres` is computed correctly -- e.g., if it's above a segment vs if it's outside the segment etc.

File: hnn_core/network_builder.py [hunk 17]
```diff
@@ -552,12 +568,19 @@ def get_data_from_neuron(self):
             isoma_py[gid] = {key: rec_i.to_python()
                              for key, rec_i in rec_i.items()}
 
+        lfp_py = list()
+        for elec in self._lfp:
+            # elec.pc.allreduce(elec.lfp_v, 1)
+            # lfp_py.append(elec_list[pos].lfp_t.to_python())
```
Comment (blakecaldwell) [inline hnn_core/network_builder.py:574]: Maybe you meant to remove these? Also, realize that you may have been testing out `allreduce` where my comment above about efficiency. Please ignore if that's the case. Main point is just that commented lines could be removed.

**Comment by jasmainak** [comment]
@ntolley do you have any local changes here? Can I work on this PR if you're busy with other PRs?

**Comment by ntolley** [comment]
No local changes! Mostly paused this to focus on the connectivity PR's. I think the last item on the list was adding tests to ensure the LFP calculation was correct? My initial thought would be to instantiate a point neuron and/or cylinder with nrn.

**Comment by cjayb** [comment]
@jasmainak are you working on this? Looks great! I'd like to propose the following for moving this ahead (and creating a test)

* use new connectivity API to prune all connections (non-connected cells, no spiking, no IPSPs)
* send (weak) synchronous EPSPs to L5 distal apical dendrites (for test: 3 x 3 grid or even a single cell would suffice)
* plot LFP as a function of depth near the cell (should go from negative to positive at soma)
* calculate far-field LFP (~10x distal-to-apical distance), which should reproduce a dipole field (see attached, calculated analytically, could be used in a test)

Screenshot 2021-04-26 at 17 29 56

**Comment by jasmainak** [comment]
I don't think I'll get to this in the coming week. If you or @ntolley want to take over, please feel free!

**Comment by cjayb** [comment]
Hmm, it's not clear to me how we can proceed to calculate LFP without implementing physical dimensions on `x` and `y`, in addition to the current `z`? Relative distances appear in the equations for "transfer resistance". Has anyone considered this? I believe NetPyNE moves the HNN network cells around to enable LFP calculations. What's the principled way of doing it, what's the "lateral extent" of the HNN network?

**Comment by cjayb** [comment]
Are our in-plane coordinates actually in physical (micron) units?! I just assumed they weren't I guess... There's something weird going on with the `psa` distance calculation, but I can't quite put my finger on it... I think I need to sit down and follow the construction of the neuron sections.

**Comment by ntolley** [comment]
> Are our in-plane coordinates actually in physical (micron) units?! I

I see why in plane units may be arbitrary for calculating the current dipole, but HNN modeling has previously published with LFP/CSD analysis: https://www.pnas.org/content/113/33/E4885

@stephanie-r-jones @samnemo do you know if there are real units for the x/y position of cells?

**Comment by samnemo** [comment]
As far as I know x,y position of cells in original model were in 'arbitrary' units primarily used to scale synaptic weights as a function of  distance. So original x,y positions should not be considered to be in microns.
________________________________

**Comment by samnemo** [comment]
Data from that paper used laminar LFP/CSD but afaik not the model. Also, NEURON internally has coordinates for cells but that doesn't mean those positions influence the circuit dynamics.
________________________________

File: hnn_core/network.py [hunk 18]
```diff
@@ -1035,6 +1043,35 @@ def clear_connectivity(self):
...
         self.connectivity = list()
 
+    def add_electrode(self, electrode_pos, sigma=3.0, method='psa'):
+        """Specify coordinates of electrodes for LFP recording.
+
+        Parameters
+        ----------
+        electrode_pos : tuple | list of tuple
+            Coordinates specifying the position for LFP electrodes in
+            the form of (x, y, z).
```
Comment (jasmainak) [inline hnn_core/network.py:1053]: I think @cjayb is right that we should specify the units here. Should be micrometer?

File: examples/plot_simulate_lfp.py [hunk 19]
```diff
@@ -0,0 +1,64 @@
...
+# Now that our electrodes are specified, we can run the simulation. The LFP
+# recordings are also stored under ``Network.lfp``.
+import matplotlib.pyplot as plt
+
+dpl = simulate_dipole(net)
+times = dpl[0].times[:-1]
+plt.figure()
+trial_idx = 0
+plt.plot(times, net.lfp[0]['data'][trial_idx])
+plt.plot(times, net.lfp[1]['data'][trial_idx])
```
Comment (jasmainak) [inline examples/plot_simulate_lfp.py:62]: add units in label ... millivolt?

**Comment by jasmainak** [comment]
> So original x,y positions should not be considered to be in microns.

humm ... okay then I'm confused too. If the x, y positions are not in microns, doesn't that mean that the LFP calculation will be in the wrong units/scale? Maybe the influence of other cells is negligible because they are "far enough" ... ?

**Comment by samnemo** [comment]
1.  original x,y positions were used for wiring the neurons in the circuit via synapses (distances were used to set snypatic delays and weights and nothing else in the ORIGINAL model afaik)
  2.  LFP was never released to public through HNN GUI - needs some work as you're all noticing - part of which involves making sure units are consistent

________________________________

File: hnn_core/network.py (hunk 18 above)
Comment (cjayb) [reply hnn_core/network.py:1053]: Also `(x, y, z)` is ambiguous: NEURON objects are `(x, z, y)` in my mind (apical dendrite oriented along z)

**Comment by cjayb** [comment]
Thanks @samnemo . It's more than "just" ensuring consistency. We have to _define_ the physical distance between cells within the "HNN column". Given that our "cells" actually represent quite large assemblies, we should probably not use biological cell-to-cell distances (on the order of microns), but at least an order of magnitude or two larger. We could base it on (Murakami & Okada, 2015)'s 1-2 nAm/mm2 for maximal neocortical tissue dipole moment. I haven't done the math yet to figure out what that implies for us, assuming the usual 10 nAm dipole moment said to be detectable using MEG.

I really like the inclusion of LFP-like signals into `hnn-core`, but these should not be confused with those obtainable in more realistic models. I'm struggling to get the `psa`-method to produce the expected dipolar field patterns, though. I need to backtrack the `imem`-currents coming from NEURON. It's probably "just" a matter of getting the distance calculation right, since the target equation is super-simple:  , where n are the sections. But we will need some physical units for all our coordinates, and to keep track of the xyz-coordinates in various places.

(Murakami & Okada, 2015): doi:10.1016/j.neuroimage.2015.02.003

**Comment by jasmainak** [comment]
I think we should just use physical distances based on known numbers from microscopic measurements.

Note that if you get something less than 1 nAm / mm^2, it could simply mean that the cells are not firing synchronously. So, I'm not sure you can calculate backwards the cell density from the signal amplitude. The Okada constant is a *maximum*. Rather, given a certain signal magnitude from our simulated signals, you can say the signal must be due to x neurons (50,000 for 10 nAm?) firing synchronously. This means that a patch of the cortex that is *at least* (x / cell density) mm^2 is activated. Interestingly, you can come at the same conclusions from the "scaling factor" in HNN.

**Comment by cjayb** [comment]
Quick shoutout here: I _think_ I've found a bug in the current implementation. I need to investigate further, but will return with more soonest. I've already taken this branch and rebased against master, hope to have an analysis (and PR) ready by Tuesday.

**Comment by jasmainak** [comment]
closed by #329
